include/boost/json/basic_parser_impl.hpp

```diff
@@ -1353,20 +1358,53 @@ parse_escaped(
             }
             if(BOOST_JSON_UNLIKELY(u1 > 0xdbff))
             {
-                BOOST_STATIC_CONSTEXPR source_location loc
-                    = BOOST_CURRENT_LOCATION;
-                return fail(cs.begin(), error::illegal_leading_surrogate,
-                    &loc);
+                // If it's an illegal leading surrogate and
+                // the parser does not allow it, return an error.
+                if(!allow_bad_utf16)
+                {
+                    BOOST_STATIC_CONSTEXPR source_location loc
+                        = BOOST_CURRENT_LOCATION;
+                    return fail(cs.begin(), error::illegal_leading_surrogate,
+                        &loc);
+                }
+                // Otherwise, append the Unicode replacement character and
+                // mark the surrogate as replaced.
+                else
+                {
+                    unsigned cp = 0xFFFD; // Unicode replacement character
+                    temp.append_utf8(cp);
+                    replaced_bad_utf16 = true;
+                }
             }
             cs += 5;
             // KRYSTIAN TODO: this can be a two byte load
             // and a single comparison. We lose error information,
             // but it's faster.
             if(BOOST_JSON_UNLIKELY(*cs != '\\'))
             {
-                BOOST_STATIC_CONSTEXPR source_location loc
-                    = BOOST_CURRENT_LOCATION;
-                return fail(cs.begin(), error::syntax, &loc);
+                // If the next character is not a backslash and
+                // the parser does not allow it, return a syntax error.
+                if(!allow_bad_utf16)
+                {
+                    BOOST_STATIC_CONSTEXPR source_location loc
+                        = BOOST_CURRENT_LOCATION;
+                    return fail(cs.begin(), error::syntax, &loc);
+                }
+                // Otherwise, append the Unicode replacement character if
+                // the first code point is a valid leading surrogate.
+                else
+                {
+                    if(replaced_bad_utf16 == false)
```

Comment:
As previously mentioned, if a bad surrogate has been encountered, at this point you should have already been broken out of the switch. Instead, here you should be _replacing_ the contents of `temp` with UTF-8-encoded `0xFFFD`, because, if you are here, than the previous surrogate was good, but now you know it's not a part of a pair.